File: src/core/spread.cpp
/*
 * Copyright 2017 Yuji Ito <llamerada.jp@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "spread.hpp"

#include <cassert>

#include "command_manager.hpp"
#include "coord_system.hpp"
#include "definition.hpp"
#include "logger.hpp"
#include "node_id.hpp"
#include "random.hpp"
#include "scheduler.hpp"
#include "utils.hpp"
#include "value_impl.hpp"

namespace colonio {
SpreadDelegate::~SpreadDelegate() {
}

Spread::Spread(
    Logger& l, Random& r, Scheduler& s, CommandManager& c, const NodeID& n, CoordSystem& cs, SpreadDelegate& d,
    const picojson::object& config) :
    CONF_CACHE_TIME(Utils::get_json<double>(config, "cacheTime", SPREAD_CACHE_TIME)),
    logger(l),
    random(r),
    scheduler(s),
    command_manager(c),
    local_nid(n),
    coord_system(cs),
    delegate(d) {
  scheduler.repeat_task(this, std::bind(&Spread::clear_cache, this), 1000);

  command_manager.set_handler(
      proto::PacketContent::kSpread, std::bind(&Spread::recv_spread, this, std::placeholders::_1));
  command_manager.set_handler(
      proto::PacketContent::kSpreadKnock, std::bind(&Spread::recv_knock, this, std::placeholders::_1));
  command_manager.set_handler(
      proto::PacketContent::kSpreadRelay, std::bind(&Spread::recv_relay, this, std::placeholders::_1));
}

Spread::~Spread() {
  scheduler.remove(this);
}

void Spread::post(
    double x, double y, double r, const std::string& name, const Value& message, uint32_t opt,
    std::function<void()>&& on_success, std::function<void(const Error&)>&& on_failure) {
  uint64_t uid              = assign_uid();
  Coordinate local_position = coord_system.get_local_position();
  Cache& c                  = cache[uid];
  c.src                     = local_nid;
  c.name                    = name;
  c.center                  = Coordinate(x, y);
  c.r                       = r;
  c.uid                     = uid;
  c.create_time             = Utils::get_current_msec();
  c.message                 = message;
  c.opt                     = opt;

  if (coord_system.get_distance(c.center, local_position) < r) {
    if (c.message.get_type() == Value::STRING_T || c.message.get_type() == Value::BINARY_T) {
      send_knock(NodeID::NONE, c);

    } else {
      for (auto& it : next_positions) {
        if (coord_system.get_distance(c.center, it.second) < r) {
          send_spread(it.first, c);
        }
      }
    }
    on_success();

  } else {
    send_relay(c, on_success, on_failure);
  }
}

void Spread::set_handler(const std::string& name, std::function<void(const Colonio::SpreadRequest&)>&& handler) {
  if (handlers.find(name) == handlers.end()) {
    handlers.insert(std::make_pair(name, handler));
  } else {
    handlers.at(name) = handler;
  }
}

void Spread::unset_handler(const std::string& name) {
  handlers.erase(name);
}

void Spread::update_next_positions(const std::map<NodeID, Coordinate>& positions) {
  next_positions = positions;
}

Spread::CommandKnock::CommandKnock(Spread& p, uint64_t u) :
    Command(PacketMode::NONE), logger(p.logger), parent(p), uid(u) {
}

void Spread::CommandKnock::on_error(ErrorCode code, const std::string& message) {
  // @todo fixme
  assert(false);
}

void Spread::CommandKnock::on_response(const Packet& packet) {
  const proto::PacketContent& c = packet.content->as_proto();
  if (!c.has_spread_response()) {
    log_debug("an unexpected packet received").map("packet", packet);
    return;
  }

  const proto::SpreadResponse& response = c.spread_response();
  // The node send response may be out of target range.
  if (!response.success()) {
    return;
  }

  auto it_c = parent.cache.find(uid);
  if (it_c != parent.cache.end()) {
    parent.send_spread(packet.src_nid, it_c->second);
  }
}

Spread::CommandRelay::CommandRelay(Spread& p, std::function<void()>& s, std::function<void(const Error&)>& f) :
    Command(PacketMode::NONE), logger(p.logger), parent(p), cb_on_success(s), cb_on_failure(f) {
}

void Spread::CommandRelay::on_response(const Packet& packet) {
  const proto::PacketContent& c = packet.content->as_proto();
  if (!c.has_spread_response()) {
    log_debug("an unexpected packet received").map("packet", packet);
    return;
  }

  const proto::SpreadResponse& response = c.spread_response();
  if (!response.success()) {
    cb_on_failure(colonio_error(static_cast<ErrorCode>(response.code()), response.message()));
    return;
  }

  cb_on_success();
}

void Spread::CommandRelay::on_error(ErrorCode code, const std::string& message) {
  // @todo output log.
  cb_on_failure(colonio_error(code, message));
}

uint64_t Spread::assign_uid() {
  uint64_t uid = random.generate_u64();
  while (cache.find(uid) != cache.end()) {
    uid = random.generate_u64();
  }
  return uid;
}

void Spread::clear_cache() {
  auto it_c = cache.begin();
  while (it_c != cache.end()) {
    if (it_c->second.create_time + CONF_CACHE_TIME < Utils::get_current_msec()) {
      it_c = cache.erase(it_c);

    } else {
      it_c++;
    }
  }
}

void Spread::recv_spread(const Packet& packet) {
  proto::Spread content = packet.content->as_proto().spread();
  Coordinate center     = Coordinate::from_pb(content.center());
  double r              = content.r();
  uint64_t uid          = content.uid();

  if (cache.find(uid) == cache.end() && coord_system.get_distance(coord_system.get_local_position(), center) < r) {
    const std::string& name = content.name();
    Cache& c                = cache[uid];
    c.src                   = NodeID::from_pb(content.source());
    c.name                  = name;
    c.center                = center;
    c.r                     = r;
    c.uid                   = uid;
    c.create_time           = Utils::get_current_msec();
    c.message               = ValueImpl::from_pb(content.message());
    c.opt                   = content.opt();

    if (c.message.get_type() == Value::STRING_T || c.message.get_type() == Value::BINARY_T) {
      send_knock(packet.src_nid, c);

    } else {
      for (auto& it_np : next_positions) {
        if (it_np.first != packet.src_nid && coord_system.get_distance(c.center, it_np.second) < r) {
          send_spread(it_np.first, c);
        }
      }
    }

    auto handler = handlers.find(name);
    if (handler != handlers.end()) {
      Colonio::SpreadRequest request;
      request.source_nid = c.src.to_str();
      request.message    = c.message;
      request.options    = c.opt;
      handler->second(request);
    }
  }
}

void Spread::recv_knock(const Packet& packet) {
  proto::SpreadKnock content = packet.content->as_proto().spread_knock();
  Coordinate center          = Coordinate::from_pb(content.center());
  double r                   = content.r();
  uint64_t uid               = content.uid();

  if (cache.find(uid) == cache.end() && coord_system.get_distance(coord_system.get_local_position(), center) < r) {
    send_response(packet, true, ErrorCode::UNDEFINED, "");
  } else {
    send_response(packet, false, ErrorCode::UNDEFINED, "");
  }
}

void Spread::recv_relay(const Packet& packet) {
  proto::SpreadRelay content = packet.content->as_proto().spread_relay();
  Coordinate center          = Coordinate::from_pb(content.center());
  uint64_t uid               = content.uid();
  uint32_t opt               = content.opt();

  if (cache.find(uid) == cache.end()) {
    Cache& c      = cache[uid];
    c.src         = NodeID::from_pb(content.source());
    c.name        = content.name();
    c.center      = center;
    c.r           = content.r();
    c.uid         = uid;
    c.create_time = Utils::get_current_msec();
    c.message     = ValueImpl::from_pb(content.message());
    c.opt         = opt;

    if (coord_system.get_distance(center, coord_system.get_local_position()) < c.r) {
      if (c.message.get_type() == Value::STRING_T || c.message.get_type() == Value::BINARY_T) {
        // @todo send_success after check the result.
        send_knock(NodeID::NONE, c);

      } else {
        for (auto& it : next_positions) {
          if (coord_system.get_distance(c.center, it.second) < c.r) {
            send_spread(it.first, c);
          }
        }
      }
      send_response(packet, true, ErrorCode::UNDEFINED, "");

      auto handler = handlers.find(c.name);
      if (handler != handlers.end()) {
        Colonio::SpreadRequest request;
        request.source_nid = c.src.to_str();
        request.message    = c.message;
        request.options    = c.opt;
        handler->second(request);
      }
      return;

    } else {
      const NodeID& dst = delegate.spread_do_get_relay_nid(center);
      if (dst == NodeID::THIS) {
        if (opt & Colonio::SPREAD_SOMEONE_MUST_RECEIVE) {
          send_response(packet, false, ErrorCode::SPREAD_NO_ONE_RECEIVE, "no one receive the message");

        } else {
          send_response(packet, true, ErrorCode::UNDEFINED, "");
        }
      } else {
        command_manager.relay_packet(dst, packet.make_copy());
      }
    }

  } else {
    if (opt & Colonio::SPREAD_SOMEONE_MUST_RECEIVE) {
      send_response(packet, false, ErrorCode::SPREAD_NO_ONE_RECEIVE, "no one receive the message");

    } else {
      send_response(packet, true, ErrorCode::UNDEFINED, "");
    }
  }
}

void Spread::send_spread(const NodeID& dst_nid, const Cache& cache) {
  std::unique_ptr<proto::PacketContent> content = std::make_unique<proto::PacketContent>();
  proto::Spread& param                          = *content->mutable_spread();

  cache.src.to_pb(param.mutable_source());
  cache.center.to_pb(param.mutable_center());
  param.set_r(cache.r);
  param.set_uid(cache.uid);
  param.set_name(cache.name);
  ValueImpl::to_pb(param.mutable_message(), cache.message);
  param.set_opt(cache.opt);

  command_manager.send_packet_one_way(dst_nid, 0, std::move(content));
}

void Spread::send_knock(const NodeID& exclude, const Cache& cache) {
  std::unique_ptr<proto::PacketContent> content = std::make_unique<proto::PacketContent>();
  proto::SpreadKnock& param                     = *content->mutable_spread_knock();

  cache.center.to_pb(param.mutable_center());
  param.set_r(cache.r);
  param.set_uid(cache.uid);

  std::shared_ptr<Packet::Content> pc = std::make_shared<Packet::Content>(std::move(content));
  for (auto& it_np : next_positions) {
    const NodeID& nid    = it_np.first;
    Coordinate& position = it_np.second;

    if (nid != exclude && coord_system.get_distance(cache.center, position) < cache.r) {
      std::shared_ptr<Command> command = std::make_shared<CommandKnock>(*this, cache.uid);
      command_manager.send_packet(nid, pc, command);
    }
  }
}

void Spread::send_relay(
    const Cache& cache, std::function<void()> on_success, std::function<void(const Error&)> on_failure) {
  std::unique_ptr<proto::PacketContent> content = std::make_unique<proto::PacketContent>();
  proto::SpreadRelay& param                     = *content->mutable_spread_relay();

  cache.src.to_pb(param.mutable_source());
  cache.center.to_pb(param.mutable_center());
  param.set_r(cache.r);
  param.set_uid(cache.uid);
  param.set_name(cache.name);
  ValueImpl::to_pb(param.mutable_message(), cache.message);
  param.set_opt(cache.opt);

  std::shared_ptr<Command> command = std::make_shared<CommandRelay>(*this, on_success, on_failure);
  const NodeID& nid                = delegate.spread_do_get_relay_nid(cache.center);
  command_manager.send_packet(nid, std::move(content), command);
}

void Spread::send_response(const Packet& packet, bool success, ErrorCode code, const std::string& message) {
  std::unique_ptr<proto::PacketContent> content = std::make_unique<proto::PacketContent>();
  proto::SpreadResponse& param                  = *content->mutable_spread_response();

  param.set_success(success);
  if (!success) {
    param.set_code(static_cast<uint32_t>(code));
    param.set_message(message);
  }

  command_manager.send_response(packet, std::move(content));
}

}  // namespace colonio

File: internal/spread/spread.go
/*
 * Copyright 2017- Yuji Ito <llamerada.jp@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package spread

import (
	"context"
	"fmt"
	"log/slog"
	"math/rand"
	"sync"
	"time"

	"github.com/llamerada-jp/colonio/config"
	"github.com/llamerada-jp/colonio/internal/constants"
	"github.com/llamerada-jp/colonio/internal/geometry"
	"github.com/llamerada-jp/colonio/internal/network/transferer"
	"github.com/llamerada-jp/colonio/internal/shared"
	"github.com/llamerada-jp/colonio/internal/wait_any"
	"github.com/llamerada-jp/colonio/proto"
)

const (
	optSomeoneMustExists = uint32(0x1)
)

var (
	ErrSomeoneMustReceiveViolation = fmt.Errorf("receiving node not found")
)

type Options struct {
	SomeoneMustExists bool
}

type Request struct {
	SourceNodeID *shared.NodeID
	Message      []byte
	Options      *Options
}

type Handler interface {
	SpreadGetRelayNodeID(position *geometry.Coordinate) *shared.NodeID
}

type cache struct {
	srcNodeID *shared.NodeID
	uid       uint64
	center    *geometry.Coordinate
	r         float64
	name      string
	message   []byte
	opt       *Options
	timestamp time.Time
}

type Spread struct {
	ctx      context.Context
	logger   *slog.Logger
	handler  Handler
	mtx      sync.RWMutex
	handlers map[string]func(*Request)

	transferer *transferer.Transferer
	config     *config.Spread
	cache      map[uint64]*cache

	localNodeID       *shared.NodeID
	coordinateSystem  geometry.CoordinateSystem
	localPosition     *geometry.Coordinate
	nextNodePositions map[shared.NodeID]*geometry.Coordinate
}

func convertOptToProto(opt *Options) uint32 {
	r := uint32(0)

	if opt.SomeoneMustExists {
		r |= optSomeoneMustExists
	}

	return r
}

func convertOptFromProto(opt uint32) *Options {
	r := &Options{}

	if (optSomeoneMustExists & opt) != 0 {
		r.SomeoneMustExists = true
	}

	return r
}

func NewSpread(ctx context.Context, logger *slog.Logger, handler Handler) *Spread {
	return &Spread{
		ctx:      ctx,
		logger:   logger,
		handler:  handler,
		handlers: make(map[string]func(*Request)),
		cache:    make(map[uint64]*cache),
	}
}

func (s *Spread) ApplyConfig(tr *transferer.Transferer, config *config.Spread, localNodeID *shared.NodeID, cs geometry.CoordinateSystem) {
	s.mtx.Lock()
	s.transferer = tr
	s.config = config
	s.localNodeID = localNodeID
	s.coordinateSystem = cs
	s.mtx.Unlock()

	transferer.SetRequestHandler[proto.PacketContent_SpreadKnock](s.transferer, s.recvKnock)
	transferer.SetRequestHandler[proto.PacketContent_SpreadRelay](s.transferer, s.recvRelay)
	transferer.SetRequestHandler[proto.PacketContent_Spread](s.transferer, s.recvSpread)

	go func() {
		ticker := time.NewTicker(config.CacheLifetime / 2)
		defer ticker.Stop()
		for {
			select {
			case <-s.ctx.Done():
				return

			case <-ticker.C:
				s.cleanupCache()
			}
		}
	}()
}

func (s *Spread) UpdateLocalPosition(pos *geometry.Coordinate) {
	s.mtx.Lock()
	defer s.mtx.Unlock()
	s.localPosition = pos
}

func (s *Spread) UpdateNextNodePosition(positions map[shared.NodeID]*geometry.Coordinate) {
	s.mtx.Lock()
	defer s.mtx.Unlock()
	s.nextNodePositions = positions
}

func (s *Spread) Post(center *geometry.Coordinate, r float64, name string, message []byte, opt *Options) error {
	s.mtx.RLock()
	defer s.mtx.RUnlock()
	tr := s.transferer

	if tr == nil {
		return fmt.Errorf("this function can be used when online and enabled Spread feature")
	}

	cache := s.makeCache(center, r, name, message, opt)

	if opt.SomeoneMustExists {
		wa := wait_any.NewWaitAny()
		if s.coordinateSystem.GetDistance(center, s.localPosition) <= r {
			for nodeID, position := range s.nextNodePositions {
				if s.coordinateSystem.GetDistance(center, position) <= r {
					wa.Add(1)
					s.sendRelay(wa, &nodeID, cache)
				}
			}

		} else {
			nextNodeID := s.handler.SpreadGetRelayNodeID(center)
			if *nextNodeID == shared.NodeIDThis {
				return ErrSomeoneMustReceiveViolation
			}
			wa.Add(1)
			s.sendRelay(wa, nextNodeID, cache)
		}

		if !wa.Wait() {
			return ErrSomeoneMustReceiveViolation
		}
		return nil
	}

	if s.coordinateSystem.GetDistance(center, s.localPosition) > r {
		nextNodeID := s.handler.SpreadGetRelayNodeID(center)
		s.sendRelay(nil, nextNodeID, cache)
		return nil
	}

	if len(message) > int(s.config.SizeToUseKnock) {
		s.sendKnock(nil, cache)

	} else {
		for nodeID, position := range s.nextNodePositions {
			if s.coordinateSystem.GetDistance(cache.center, position) < r {
				s.sendSpread(&nodeID, cache)
			}
		}
	}
	return nil
}

func (s *Spread) SetHandler(name string, handler func(*Request)) {
	s.mtx.Lock()
	defer s.mtx.Unlock()

	if _, ok := s.handlers[name]; ok {
		s.logger.Warn("handler already exists", slog.String("name", name))
	}

	s.handlers[name] = handler
}

func (s *Spread) UnsetHandler(name string) {
	s.mtx.Lock()
	defer s.mtx.Unlock()

	if _, ok := s.handlers[name]; !ok {
		s.logger.Warn("handler not found", slog.String("name", name))
	}

	delete(s.handlers, name)
}

func (s *Spread) makeCache(center *geometry.Coordinate, r float64, name string, message []byte, opt *Options) *cache {
	uid := rand.Uint64()
	for uid == 0 || s.cache[uid] != nil {
		uid = rand.Uint64()
	}

	cache := &cache{
		srcNodeID: s.localNodeID,
		uid:       uid,
		center:    center,
		r:         r,
		name:      name,
		message:   message,
		opt:       opt,
		timestamp: time.Now(),
	}

	s.cache[uid] = cache
	return cache
}

func (s *Spread) makeCacheWithUID(srcNodeID *shared.NodeID, uid uint64, center *geometry.Coordinate, r float64, name string, message []byte, opt *Options) *cache {
	cache := &cache{
		srcNodeID: srcNodeID,
		uid:       uid,
		center:    center,
		r:         r,
		name:      name,
		message:   message,
		opt:       opt,
		timestamp: time.Now(),
	}

	s.cache[uid] = cache
	return cache
}

func (s *Spread) recvKnock(packet *shared.Packet) {
	content := packet.Content.GetSpreadKnock()
	center := geometry.NewCoordinateFromProto(content.Center)
	r := content.R
	uid := content.Uid

	s.mtx.RLock()
	defer s.mtx.RUnlock()

	accept := false
	if _, ok := s.cache[uid]; !ok && s.coordinateSystem.GetDistance(center, s.localPosition) <= r {
		accept = true
	}

	response := &proto.PacketContent{
		Content: &proto.PacketContent_SpreadKnockResponse{
			SpreadKnockResponse: &proto.SpreadKnockResponse{
				Accept: accept,
			},
		},
	}
	s.transferer.Response(packet, response)
}

func (s *Spread) recvRelay(packet *shared.Packet) {
	content := packet.Content.GetSpreadRelay()
	uid := content.Uid
	oneWay := (packet.Mode & shared.PacketModeOneWay) != 0
	center := geometry.NewCoordinateFromProto(content.Center)
	r := content.R
	name := content.Name

	s.mtx.Lock()
	defer s.mtx.Unlock()

	if _, ok := s.cache[uid]; ok {
		if !oneWay {
			s.responseForRelay(packet, false)
		}
		return
	}

	cache := s.makeCacheWithUID(
		shared.NewNodeIDFromProto(content.Source),
		uid, center, r, name,
		content.Message,
		convertOptFromProto(content.Opt))

	if s.coordinateSystem.GetDistance(center, s.localPosition) > r {
		nextNodeID := s.handler.SpreadGetRelayNodeID(center)
		if *nextNodeID == shared.NodeIDThis {
			if !oneWay {
				s.responseForRelay(packet, false)
			}
			return
		}
		s.transferer.Relay(nextNodeID, packet)
		return
	}

	handler := s.handlers[name]
	if handler != nil {
		go handler(&Request{
			SourceNodeID: cache.srcNodeID,
			Message:      cache.message,
			Options:      cache.opt,
		})
	}

	if !oneWay {
		s.responseForRelay(packet, true)
	}

	if len(cache.message) > int(s.config.SizeToUseKnock) {
		s.sendKnock(packet.SrcNodeID, cache)

	} else {
		for nodeID, position := range s.nextNodePositions {
			if s.coordinateSystem.GetDistance(cache.center, position) < r {
				s.sendSpread(&nodeID, cache)
			}
		}
	}
}

func (s *Spread) responseForRelay(packet *shared.Packet, success bool) {
	response := &proto.PacketContent{
		Content: &proto.PacketContent_SpreadRelayResponse{
			SpreadRelayResponse: &proto.SpreadRelayResponse{
				Success: success,
			},
		},
	}
	s.transferer.Response(packet, response)
}

func (s *Spread) recvSpread(packet *shared.Packet) {
	content := packet.Content.GetSpread()
	center := geometry.NewCoordinateFromProto(content.Center)
	r := content.R
	uid := content.Uid

	s.mtx.Lock()
	defer s.mtx.Unlock()

	if _, ok := s.cache[uid]; ok || s.coordinateSystem.GetDistance(center, s.localPosition) > r {
		return
	}

	cache := s.makeCacheWithUID(
		shared.NewNodeIDFromProto(content.Source),
		uid, center, r,
		content.Name,
		content.Message,
		convertOptFromProto(content.Opt))

	handler := s.handlers[content.Name]
	if handler != nil {
		go handler(&Request{
			SourceNodeID: cache.srcNodeID,
			Message:      cache.message,
			Options:      cache.opt,
		})
	}

	if len(cache.message) > int(s.config.SizeToUseKnock) {
		s.sendKnock(packet.SrcNodeID, cache)

	} else {
		for nodeID, position := range s.nextNodePositions {
			if *packet.SrcNodeID != nodeID && s.coordinateSystem.GetDistance(cache.center, position) < r {
				s.sendSpread(&nodeID, cache)
			}
		}
	}
}

func (s *Spread) sendKnock(exclude *shared.NodeID, cache *cache) {
	content := &proto.PacketContent{
		Content: &proto.PacketContent_SpreadKnock{
			SpreadKnock: &proto.SpreadKnock{
				Center: cache.center.Proto(),
				R:      cache.r,
				Uid:    cache.uid,
			},
		},
	}

	for nodeID, position := range s.nextNodePositions {
		if exclude != nil && *exclude == nodeID {
			continue
		}

		handler := &knockHandler{
			logger:    s.logger,
			cache:     cache,
			dstNodeID: &nodeID,
			sender:    s.sendSpread,
		}

		if s.coordinateSystem.GetDistance(cache.center, position) < cache.r {
			s.transferer.Request(&nodeID, shared.PacketModeNone, content, handler)
		}
	}
}

type knockHandler struct {
	logger    *slog.Logger
	cache     *cache
	dstNodeID *shared.NodeID
	sender    func(dstNodeID *shared.NodeID, cache *cache)
}

func (h *knockHandler) OnResponse(p *shared.Packet) {
	response := p.Content.GetSpreadKnockResponse()
	if response == nil {
		h.logger.Warn("invalid packet received")
		return
	}

	if !response.Accept {
		return
	}

	h.sender(h.dstNodeID, h.cache)
}

func (h *knockHandler) OnError(code constants.PacketErrorCode, message string) {
	h.logger.Warn(message, slog.Int("code", int(code)))
}

func (s *Spread) sendRelay(wa *wait_any.WaitAny, dstNodeID *shared.NodeID, cache *cache) {
	content := &proto.PacketContent{
		Content: &proto.PacketContent_SpreadRelay{
			SpreadRelay: &proto.SpreadRelay{
				Source:  cache.srcNodeID.Proto(),
				Center:  cache.center.Proto(),
				R:       cache.r,
				Uid:     cache.uid,
				Name:    cache.name,
				Message: cache.message,
				Opt:     convertOptToProto(cache.opt),
			},
		},
	}

	if cache.opt.SomeoneMustExists {
		handler := &relayHandler{
			logger: s.logger,
			wa:     wa,
		}
		s.transferer.Request(dstNodeID, shared.PacketModeNone, content, handler)

	} else {
		s.transferer.RequestOneWay(dstNodeID, shared.PacketModeNone, content)
	}
}

func (s *Spread) sendSpread(dstNodeID *shared.NodeID, cache *cache) {
	content := &proto.PacketContent{
		Content: &proto.PacketContent_Spread{
			Spread: &proto.Spread{
				Source:  cache.srcNodeID.Proto(),
				Center:  cache.center.Proto(),
				R:       cache.r,
				Uid:     cache.uid,
				Name:    cache.name,
				Message: cache.message,
				Opt:     convertOptToProto(cache.opt),
			},
		},
	}

	s.transferer.RequestOneWay(dstNodeID, shared.PacketModeNone, content)
}

type relayHandler struct {
	logger *slog.Logger
	wa     *wait_any.WaitAny
}

func (h *relayHandler) OnResponse(p *shared.Packet) {
	response := p.Content.GetSpreadRelayResponse()
	if response == nil {
		h.logger.Warn("invalid packet received")
		h.wa.Done(false)
		return
	}

	h.wa.Done(response.Success)
}

func (h *relayHandler) OnError(code constants.PacketErrorCode, message string) {
	h.logger.Warn(message, slog.Int("code", int(code)))
	h.wa.Done(false)
}

func (s *Spread) cleanupCache() {
	s.mtx.Lock()
	defer s.mtx.Unlock()

	now := time.Now()
	for uid, cache := range s.cache {
		if now.After(cache.timestamp.Add(s.config.CacheLifetime)) {
			delete(s.cache, uid)
		}
	}
}
